src/python/nxt_python.c

```diff
@@ -403,11 +403,13 @@ nxt_python_set_target(nxt_task_t *task, nxt_python_target_t *target,
     char              *callable, *module_name;
     PyObject          *module, *obj;
     nxt_str_t         str;
+    nxt_bool_t        is_factory = 0;
```

Comment:
That test is just to see if the factory _variable_ is set. It may be set to `false`.

Reply:
If you mean something like

```c
    value = nxt_conf_get_object_member(conf, &factory_flag_str, NULL);          
    if (!value) {                                                        
        if (nxt_slow_path(PyCallable_Check(obj) == 0)) {                        
            nxt_alert(task, "\"%s\" in module \"%s\" is not a callable object", 
                      callable, module_name);                                   
            goto fail;                                                          
        }                                                                       
    } else {                                                                    
        nxt_bool_t  is_factory = nxt_conf_get_boolean(value);
                               
        if (is_factory) {                                                       
            if (nxt_slow_path(PyCallable_Check(obj) == 0)) {                    
                nxt_alert(task,                                                 
                          "factory \"%s\" in module \"%s\" ",                   
                          "can not be called to fetch callable", callable,      
                          module_name);                                         
                goto fail;                                                      
            }                                                                   
                                                                                
            obj = PyObject_CallObject(obj, NULL);                               
            if (nxt_slow_path(PyCallable_Check(obj) == 0)) {                    
                nxt_alert(task,                                                 
                          "factory \"%s\" in module \"%s\" ",                   
                          "did not return callable object", callable, module_name);                                         
                goto fail;                                                      
            }                                                                   
        }                                                                       
    }
```

Aside from adding an extra level of indentation. If `is_factory` is `false` we still need to do the `nxt_slow_path(PyCallable_Check(obj) == 0` check... 

We also want to distinguish between a `factory` failure vs an ordinary `callable` failure and without duplicating the above `callable` check.

If on the other hand you mean something like (untested)

```c
    value = nxt_conf_get_object_member(conf, &factory_flag_str, NULL);          
    if (value != NULL && nxt_conf_get_boolean(value) == 1) {                                                    
        if (nxt_slow_path(PyCallable_Check(obj) == 0)) {                        
            nxt_alert(task,                                                     
                      "factory \"%s\" in module \"%s\" ",                       
                      "can not be called to fetch callable", callable,
                      module_name);
            goto fail;                                                          
        }                                                                       
                                                                                
        obj = PyObject_CallObject(obj, NULL);                                   
        if (nxt_slow_path(PyCallable_Check(obj) == 0)) {                        
            nxt_alert(task,                                                     
                      "factory \"%s\" in module \"%s\" ",                       
                      "did not return callable object", callable, module_name); 
            goto fail;                                                          
        }                                                                       
                                                                                
    } else if (nxt_slow_path(PyCallable_Check(obj) == 0)) {                     
        nxt_alert(task, "\"%s\" in module \"%s\" is not a callable object",     
                  callable, module_name);                                       
        goto fail;                                                              
    }
```

It's not terrible...

Reply:
Author has already added what I suggested:

```
    if (value != NULL) {
        is_factory = nxt_conf_get_boolean(value);
    } else {
        is_factory = 0;
    }
```

Thanks @gourav-kandoria

Reply:
I'm not a fan of

```c
if (value != NULL && nxt_conf_get_boolean(value) == 1) {
```

and I wouldn't want to see that kind of thing spread, but I _could_ live with it in this case...

Reply:
> Author has already added what I suggested:
> 
> ```
>     if (value != NULL) {
>         is_factory = nxt_conf_get_boolean(value);
>     } else {
>         is_factory = 0;
>     }
> ```
> 
> Thanks @gourav-kandoria

Sorry, there is no way that is an improvement.

Reply:
The thing is I explicitly asked for it to be done the way it was to avoid doing it the above way which just introduces unnecessary code.

You mention about _optimisations_. I'm curious _what_ use case you're optimising for?

This is not a hot path.

Even then I doubt the above is an optimisation as it introduces more code, which generates more assembly and will likely have a larger i-cache footprint.

To illustrate take the following exampe  

```c
int main(void)
{
	int x = 42;
	int i = 0;

	if (x == 42)
		i = x;

	return 0;
}
```

This produces the following assembly

```asm
.LFB0:
	.cfi_startproc
	pushq	%rbp
	.cfi_def_cfa_offset 16
	.cfi_offset 6, -16
	movq	%rsp, %rbp
	.cfi_def_cfa_register 6
	movl	$42, -4(%rbp)
	movl	$0, -8(%rbp)
	cmpl	$42, -4(%rbp)
	jne	.L2
	movl	-4(%rbp), %eax
	movl	%eax, -8(%rbp)
.L2:
	movl	$0, %eax
	popq	%rbp
	.cfi_def_cfa 7, 8
	ret
	.cfi_endproc
```

You can see thw two `movl` instructions initialising the variables.

We have one compare and jump.

Now lets look at what happens when we make it a if/then/else

```c
int main(void)
{
	int x = 42;
	int i;

	if (x == 42)
		i = x;
	else
		i = 0;

	return 0;
}
```

That produces the following assembly

```asm
.LFB0:
	.cfi_startproc
	pushq	%rbp
	.cfi_def_cfa_offset 16
	.cfi_offset 6, -16
	movq	%rsp, %rbp
	.cfi_def_cfa_register 6
	movl	$42, -4(%rbp)
	cmpl	$42, -4(%rbp)
	jne	.L2
	movl	-4(%rbp), %eax
	movl	%eax, -8(%rbp)
	jmp	.L3
.L2:
	movl	$0, -8(%rbp)
.L3:
	movl	$0, %eax
	popq	%rbp
	.cfi_def_cfa 7, 8
	ret
	.cfi_endproc
```

Yes, there is only one `movl` instruction at the start, but we now have two jumps.

In the common case that I would suggest is that `factory` is _not_ set, `is_factory` is set to 0 (once) in both cases.

But again this is all kind of moot, as this is not a hot path and cleaner code wins.

@gourav-kandoria 

I would _strongly_ suggest reverting this change...

Reply:
> as this is not a hot path

Then it make sense to optimise it using `nxt_slow_path()`.

Reply:
> > as this is not a hot path

By that I mean `nxt_python_set_target()` as a whole, it's only called at startup and _maybe_ reconfigure time, not thousands or even hundreds or tens of times a second.

> Then it make sense to optimise it using `nxt_slow_path()`.

We could, I'm not sue what we're optimising for here though.

likely/unlikely should really only be used in cases where you know for certain it's one way or the other a majority of time in all cases/workloads.

We seem to sprinkle these things around willy nilly.

Error branches are perhaps still a valid use... but again you really need to measure/profile this stuff...

Reply:
@ac000 sure, can revert it. BTW good to see this type of discussion. gave me some perspective about how to think about  things at this level. just before reverting. @andrey-zelenkov are you also okay with @ac000 viewpoint. would wait for your reply before making any change as I would like to make only the one and final change now 😅 😅